### ISSUE TITLE
Incorrect error for overflow in bound evaluation

### ISSUE THREAD

**Issue by wthrowe** [rust-lang/rust#29446]
``` rust
#![allow(unconditional_recursion)]

struct X<T>(T);

trait A {
    fn foo(self);
}

impl<T> A for T {
    fn foo(self) {
        X(self).foo() // <-- error points here
    }
}

fn main() {
    1.foo();
}
```

reports `error: overflow evaluating the requirement `i32 : core::marker::Sized` [E0275]`.  While this code certainly shouldn't work, I would not expect proving `i32: Sized` to be the problematic part.

(This is similar to #23714, but I think that's about the example failing to compile rather than the nonsense error.)

**Comment by arielb1** [comment]
The note in the error message correctly tells you that the problem involves the type `X<X<X<X<X<X<X<X<i32>>>>>>>>`.

**Comment by arielb1** [comment]
The error here is a recursive monomorphization (`<T as A>::foo` requiring `<X<T> as A>::foo`) that bails out in trait selection. The error message sucks.

**Comment by Mark-Simulacrum** [comment]
Today's error message is quite a bit more clear, I think, since it indicates that the problem is with this piece of code and with what expansion. As such, I'm going to go ahead and close this.

```
error: reached the recursion limit while instantiating `<T as A><X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<X<i32>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>::foo`
  --> test.rs:10:5
   |
10 | /     fn foo(self) {
11 | |         X(self).foo() // <-- error points here
12 | |     }
   | |_____^
```
